### ISSUE TITLE
AWS -Still incorrect path when using $CODEBUILD_SRC_DIR

### ISSUE THREAD

**Issue by kutscherma** [serverless/serverless#6964]
**My Situation**
I am using AWS CodeBuild and the newest Serverless Framework to build and deploy my Node.js application as a Lambda. 
I package everything with `severless package --package /PATH/..` and store the artifacts under this path in a S3.
In my deployment step I use the CODEBUILD_SRC_DIR to get the path/location of the current build artifacts which I want to deploy: 
`sls deploy --stage $env --package $CODEBUILD_SRC_DIR/target/$env -v -r eu-central-1`

**Problem**
I get the following error:
Error: ENOENT: no such file or directory, open '/codebuild/output/src067167590/src/codebuild/output/src067167590/src/target/dev/serverless-state.json' 

As you can see there seems to be something wrong with the path which I get back from $CODEBUILD_SRC_DIR . It doubles "/codebuild/output/src067167590" and therefore it cannot find the files.
Strange is when I do ` echo $CODEBUILD_SRC_DIR` everything seems to be fine. Is it a problem with the command `sls deploy`?

**My Solution**
Like the other guys mentioned I rolled also back to serverless 1.38. Then it works.

Is it sill not fixed ?

Thanks!

**Comment by medikoo** [comment]
@soelchma thanks for report. Can you show full output with `SLS_DEBUG=*`, and possibly track which exactly fs invocation crashes with `ENOENT` error?

**Comment by ferasallaou** [comment]
I've the same issue as well, Once the script enters deployment stage in Pipeline everything seems to be fine, it uploads everything and generate the URL. but then it fails due to the same error with duplicating it's directories
`'/codebuild/output/src932745260/src/core/codebuild/output/src932745260/src/core/deployment_package/serverless-state.json' `

Also, even when I specify --package flag, serverless.yml keep looking for nodeModules.zip, which is my Layer directory inside .serverless and ignores that I've packaged it inside deployment_package.

**Comment by medikoo** [comment]
@ferasallaou Can you show full output with `SLS_DEBUG=*`?

**Comment by ferasallaou** [comment]
@medikoo Here you go
```
Serverless: Load command interactiveCli 
Serverless: Load command config 
Serverless: Load command config:credentials 
Serverless: Load command config:tabcompletion 
Serverless: Load command config:tabcompletion:install 
Serverless: Load command config:tabcompletion:uninstall 
Serverless: Load command create 
Serverless: Load command install 
Serverless: Load command package 
Serverless: Load command deploy 
Serverless: Load command deploy:function 
Serverless: Load command deploy:list 
Serverless: Load command deploy:list:functions 
Serverless: Load command invoke 
Serverless: Load command invoke:local 
Serverless: Load command info 
Serverless: Load command logs 
Serverless: Load command metrics 
Serverless: Load command print 
Serverless: Load command remove 
Serverless: Load command rollback 
Serverless: Load command rollback:function 
Serverless: Load command slstats 
Serverless: Load command plugin 
Serverless: Load command plugin 
Serverless: Load command plugin:install 
Serverless: Load command plugin 
Serverless: Load command plugin:uninstall 
Serverless: Load command plugin 
Serverless: Load command plugin:list 
Serverless: Load command plugin 
Serverless: Load command plugin:search 
Serverless: Load command config 
Serverless: Load command config:credentials 
Serverless: Load command rollback 
Serverless: Load command rollback:function 
Serverless: Load command login 
Serverless: Load command logout 
Serverless: Load command generate-event 
Serverless: Load command test 
Serverless: Load command dashboard 
Serverless: Invoke deploy 
Serverless: Invoke aws:common:validate 
Serverless: Invoke aws:common:moveArtifactsToTemp 
Serverless: Invoke aws:deploy:deploy 
Serverless: [AWS cloudformation 200 0.07s 0 retries] describeStacks({ StackName: 'Feras-App-development' }) 
Serverless: [AWS cloudformation 200 0.06s 0 retries] describeStackResource({ StackName: 'Feras-App-development', 
  LogicalResourceId: 'ServerlessDeploymentBucket' }) 
Serverless: [AWS s3 200 0.075s 0 retries] listObjectsV2({ Bucket: 'feras-app-development-serverlessdeploymentbucket-10qsuhijjrcsj', 
  Prefix: 'serverless/Feras-App/development' }) 
Serverless: [AWS s3 200 0.035s 0 retries] headObject({ Bucket: 'feras-app-development-serverlessdeploymentbucket-10qsuhijjrcsj', 
  Key: 'serverless/Feras-App/development/1574071109852-2019-11-18T09:58:29.852Z/Feras-App.zip' }) 
Serverless: [AWS s3 200 0.047s 0 retries] headObject({ Bucket: 'feras-app-development-serverlessdeploymentbucket-10qsuhijjrcsj', 
  Key: 'serverless/Feras-App/development/1574071109852-2019-11-18T09:58:29.852Z/compiled-cloudformation-template.json' }) 
Serverless: [AWS lambda 200 0.093s 0 retries] getFunction({ FunctionName: 'Feras-App-development-app' }) 
Serverless: [AWS sts 200 0.325s 0 retries] getCallerIdentity({}) 
Serverless: Uploading CloudFormation file to S3... 
Serverless: [AWS s3 200 0.074s 0 retries] putObject({ Body: <Buffer 7b 22 41 57 53 54 65 6d 70 6c 61 74 65 46 6f 72 6d 61 74 56 65 72 73 69 6f 6e 22 3a 22 32 30 31 30 2d 30 39 2d 30 39 22 2c 22 44 65 73 63 72 69 70 74 ... >, 
  Bucket: 'feras-app-development-serverlessdeploymentbucket-10qsuhijjrcsj', 
  Key: 'serverless/Feras-App/development/1574094144168-2019-11-18T16:22:24.168Z/compiled-cloudformation-template.json', 
  ContentType: 'application/json', 
  Metadata: { filesha256: 'i3fMix8mBDJGLsu7iJ75dP9jNdeFYy8Zt+AmADpV8qo=' } }) 
Serverless: Uploading artifacts... 
Serverless: Uploading service Feras-App.zip file to S3 (21.28 MB)... 
Serverless: [AWS s3 200 0.121s 0 retries] createMultipartUpload({ Bucket: 'feras-app-development-serverlessdeploymentbucket-10qsuhijjrcsj', 
  Key: 'serverless/Feras-App/development/1574094144168-2019-11-18T16:22:24.168Z/Feras-App.zip', 
  ContentType: 'application/zip', 
  Metadata: { filesha256: 'RlbTMRES7wfUg02UBiCU6r++PqxHYyEXk+ku2udNVe8=' } }) 
Serverless: [AWS s3 200 0.172s 0 retries] uploadPart({ Body: <Buffer e6 a9 08 54 95 78 94 d3 48 b3 18 99 3f db 8c 6f 45 6a 09 5b 15 4c f1 c2 d3 f2 a6 75 6c 50 74 ec 8a e6 38 d0 95 73 73 c8 64 ed 56 a8 0d af 54 2b d9 80 ... >, 
  ContentLength: 5242880, 
  PartNumber: 2, 
  Bucket: 'feras-app-development-serverlessdeploymentbucket-10qsuhijjrcsj', 
  Key: 'serverless/Feras-App/development/1574094144168-2019-11-18T16:22:24.168Z/Feras-App.zip', 
  UploadId: 'ocAsuo6lsMrk9.MITA84HC3DB5MBRayxYOfivrohIAk62vBC9ePjgSkX94aT7GQ8b7YA00FftarbFxwVf3NMdhwZYIQm4fK1igD_gCO.43LHcRJxJVQ00CcTHmQ3XqjVBIOPz40gE5WO5pS.XBOt7KU.yYnOT5j.7JoTbcBhSbGdP946_6vpfjfXjFNHohnL' }) 
Serverless: [AWS s3 200 0.161s 0 retries] uploadPart({ Body: <Buffer 8a 53 65 16 b5 55 68 e6 9a d2 c8 2d dd 90 4f 2e 80 9f 7c f8 40 a3 f7 a6 56 d2 e4 c7 29 b7 f7 e5 74 49 c8 1f da 38 43 a4 6d dc b5 98 d6 ae 6c b6 c2 b6 ... >, 
  ContentLength: 5242880, 
  PartNumber: 4, 
  Bucket: 'feras-app-development-serverlessdeploymentbucket-10qsuhijjrcsj', 
  Key: 'serverless/Feras-App/development/1574094144168-2019-11-18T16:22:24.168Z/Feras-App.zip', 
  UploadId: 'ocAsuo6lsMrk9.MITA84HC3DB5MBRayxYOfivrohIAk62vBC9ePjgSkX94aT7GQ8b7YA00FftarbFxwVf3NMdhwZYIQm4fK1igD_gCO.43LHcRJxJVQ00CcTHmQ3XqjVBIOPz40gE5WO5pS.XBOt7KU.yYnOT5j.7JoTbcBhSbGdP946_6vpfjfXjFNHohnL' }) 
Serverless: [AWS s3 200 0.202s 0 retries] uploadPart({ Body: <Buffer 99 9d 27 7d b7 7f 76 1b 07 7a 3a 7b b4 f5 9c 05 5a 4c d5 d6 cc e9 94 86 e3 30 4a 82 38 38 24 34 21 83 00 68 98 84 f1 61 38 a6 03 0a d1 21 89 08 09 26 ... >, 
  ContentLength: 5242880, 
  PartNumber: 3, 
  Bucket: 'feras-app-development-serverlessdeploymentbucket-10qsuhijjrcsj', 
  Key: 'serverless/Feras-App/development/1574094144168-2019-11-18T16:22:24.168Z/Feras-App.zip', 
  UploadId: 'ocAsuo6lsMrk9.MITA84HC3DB5MBRayxYOfivrohIAk62vBC9ePjgSkX94aT7GQ8b7YA00FftarbFxwVf3NMdhwZYIQm4fK1igD_gCO.43LHcRJxJVQ00CcTHmQ3XqjVBIOPz40gE5WO5pS.XBOt7KU.yYnOT5j.7JoTbcBhSbGdP946_6vpfjfXjFNHohnL' }) 
Serverless: [AWS s3 200 0.24s 0 retries] uploadPart({ Body: <Buffer 50 4b 03 04 14 00 08 00 08 00 00 00 21 00 00 00 00 00 00 00 00 00 00 00 00 00 04 00 00 00 2e 65 6e 76 4d 4f 31 6e c3 30 10 db fd 0a 6f da 9a bd 80 07 ... >, 
  ContentLength: 5242880, 
  PartNumber: 1, 
  Bucket: 'feras-app-development-serverlessdeploymentbucket-10qsuhijjrcsj', 
  Key: 'serverless/Feras-App/development/1574094144168-2019-11-18T16:22:24.168Z/Feras-App.zip', 
  UploadId: 'ocAsuo6lsMrk9.MITA84HC3DB5MBRayxYOfivrohIAk62vBC9ePjgSkX94aT7GQ8b7YA00FftarbFxwVf3NMdhwZYIQm4fK1igD_gCO.43LHcRJxJVQ00CcTHmQ3XqjVBIOPz40gE5WO5pS.XBOt7KU.yYnOT5j.7JoTbcBhSbGdP946_6vpfjfXjFNHohnL' }) 
Serverless: [AWS s3 200 0.123s 0 retries] uploadPart({ Body: <Buffer 6d 70 6c 61 74 65 2f 6c 69 62 2f 70 61 72 73 65 2e 6a 73 50 4b 01 02 2d 03 14 00 08 00 08 00 00 00 21 00 58 c6 97 ed d5 05 00 00 39 12 00 00 2c 00 00 ... >, 
  ContentLength: 1344665, 
  PartNumber: 5, 
  Bucket: 'feras-app-development-serverlessdeploymentbucket-10qsuhijjrcsj', 
  Key: 'serverless/Feras-App/development/1574094144168-2019-11-18T16:22:24.168Z/Feras-App.zip', 
  UploadId: 'ocAsuo6lsMrk9.MITA84HC3DB5MBRayxYOfivrohIAk62vBC9ePjgSkX94aT7GQ8b7YA00FftarbFxwVf3NMdhwZYIQm4fK1igD_gCO.43LHcRJxJVQ00CcTHmQ3XqjVBIOPz40gE5WO5pS.XBOt7KU.yYnOT5j.7JoTbcBhSbGdP946_6vpfjfXjFNHohnL' }) 
Serverless: [AWS s3 200 0.128s 0 retries] completeMultipartUpload({ MultipartUpload:  
   { Parts:  
      [ { ETag: '"1f3d23f9b855b58c635ba6e2f22a2f90"', PartNumber: 1 }, 
        { ETag: '"fb87edaea54f52f8726fb06fcff874a3"', PartNumber: 2 }, 
        { ETag: '"0e2b3518bdd84dec281e93f03f622c17"', PartNumber: 3 }, 
        { ETag: '"58c79060e577e10e83a0bd8cbb0d358c"', PartNumber: 4 }, 
        { ETag: '"d0dd9e4e4a48748210d51dceffdee62a"', PartNumber: 5 }, 
        [length]: 5 ] }, 
  Bucket: 'feras-app-development-serverlessdeploymentbucket-10qsuhijjrcsj', 
  Key: 'serverless/Feras-App/development/1574094144168-2019-11-18T16:22:24.168Z/Feras-App.zip', 
  UploadId: 'ocAsuo6lsMrk9.MITA84HC3DB5MBRayxYOfivrohIAk62vBC9ePjgSkX94aT7GQ8b7YA00FftarbFxwVf3NMdhwZYIQm4fK1igD_gCO.43LHcRJxJVQ00CcTHmQ3XqjVBIOPz40gE5WO5pS.XBOt7KU.yYnOT5j.7JoTbcBhSbGdP946_6vpfjfXjFNHohnL' }) 
Serverless: Validating template... 
Serverless: [AWS cloudformation 200 0.339s 0 retries] validateTemplate({ TemplateURL: 'https://s3.amazonaws.com/feras-app-development-serverlessdeploymentbucket-10qsuhijjrcsj/serverless/Feras-App/development/1574094144168-2019-11-18T16:22:24.168Z/compiled-cloudformation-template.json' }) 
Serverless: Updating Stack... 
Serverless: [AWS cloudformation 200 0.422s 0 retries] updateStack({ StackName: 'Feras-App-development', 
  Capabilities: [ 'CAPABILITY_IAM', 'CAPABILITY_NAMED_IAM', [length]: 2 ], 
  Parameters: [ [length]: 0 ], 
  TemplateURL: 'https://s3.amazonaws.com/feras-app-development-serverlessdeploymentbucket-10qsuhijjrcsj/serverless/Feras-App/development/1574094144168-2019-11-18T16:22:24.168Z/compiled-cloudformation-template.json', 
  Tags: [ { Key: 'STAGE', Value: 'development' }, [length]: 1 ] }) 
Serverless: Checking Stack update progress... 
Serverless: [AWS cloudformation 200 0.125s 0 retries] describeStackEvents({ StackName: 'arn:aws:cloudformation:us-west-2:227403157463:stack/Feras-App-development/3f824ae0-09dc-11ea-ad58-025a30c055ca' }) 
.Serverless: [AWS cloudformation 200 0.076s 0 retries] describeStackEvents({ StackName: 'arn:aws:cloudformation:us-west-2:227403157463:stack/Feras-App-development/3f824ae0-09dc-11ea-ad58-025a30c055ca' }) 
..Serverless: [AWS cloudformation 200 0.138s 0 retries] describeStackEvents({ StackName: 'arn:aws:cloudformation:us-west-2:227403157463:stack/Feras-App-development/3f824ae0-09dc-11ea-ad58-025a30c055ca' }) 
.......Serverless: [AWS cloudformation 200 0.119s 0 retries] describeStackEvents({ StackName: 'arn:aws:cloudformation:us-west-2:227403157463:stack/Feras-App-development/3f824ae0-09dc-11ea-ad58-025a30c055ca' }) 
.... 
Serverless: Stack update finished... 
Serverless: Invoke aws:info 
Serverless: [AWS cloudformation 200 0.061s 0 retries] describeStacks({ StackName: 'Feras-App-development' }) 
Serverless: [AWS cloudformation 200 0.056s 0 retries] listStackResources({ StackName: 'Feras-App-development' }) 
Service Information 
service: Feras-App 
stage: development 
region: us-west-2 
stack: Feras-App-development 
resources: 14 
api keys: 
  None 
endpoints: 
  ANY - https://884orkrq50.execute-api.us-west-2.amazonaws.com/development/ 
  ANY - https://884orkrq50.execute-api.us-west-2.amazonaws.com/development/{proxy+} 
functions: 
  app: Feras-App-development-app 
layers: 
  None 
  
  Error -------------------------------------------------- 
  
  Error: ENOENT: no such file or directory, open '/codebuild/output/src663429830/src/core/codebuild/output/src663429830/src/core/deployment_package/serverless-state.json' 
      at Object.fs.openSync (fs.js:646:18) 
      at Object.fs.readFileSync (fs.js:551:33) 
      at readFileSync (/usr/local/lib/node_modules/serverless/lib/utils/fs/readFileSync.js:7:24) 
      at Utils.readFileSync (/usr/local/lib/node_modules/serverless/lib/classes/Utils.js:84:12) 
      at AwsCompileApigEvents.getServiceState (/usr/local/lib/node_modules/serverless/lib/plugins/aws/lib/getServiceState.js:12:34) 
      at Object.after:deploy:deploy [as hook] (/usr/local/lib/node_modules/serverless/lib/plugins/aws/package/compile/events/apiGateway/index.js:77:39) 
      at BbPromise.reduce (/usr/local/lib/node_modules/serverless/lib/classes/PluginManager.js:489:55) 
  From previous event: 
      at PluginManager.invoke (/usr/local/lib/node_modules/serverless/lib/classes/PluginManager.js:489:22) 
      at getHooks.reduce.then (/usr/local/lib/node_modules/serverless/lib/classes/PluginManager.js:524:24) 
  From previous event: 
      at PluginManager.run (/usr/local/lib/node_modules/serverless/lib/classes/PluginManager.js:524:8) 
      at variables.populateService.then (/usr/local/lib/node_modules/serverless/lib/Serverless.js:115:33) 
      at runCallback (timers.js:810:20) 
      at tryOnImmediate (timers.js:768:5) 
      at processImmediate [as _immediateCallback] (timers.js:745:5) 
  From previous event: 
      at Serverless.run (/usr/local/lib/node_modules/serverless/lib/Serverless.js:102:74) 
      at serverless.init.then (/usr/local/lib/node_modules/serverless/bin/serverless.js:72:30) 
      at /usr/local/lib/node_modules/serverless/node_modules/graceful-fs/graceful-fs.js:111:16 
      at /usr/local/lib/node_modules/serverless/node_modules/graceful-fs/graceful-fs.js:45:10 
      at FSReqWrap.oncomplete (fs.js:135:15) 
  From previous event: 
      at initializeErrorReporter.then (/usr/local/lib/node_modules/serverless/bin/serverless.js:72:8) 
      at runCallback (timers.js:810:20) 
      at tryOnImmediate (timers.js:768:5) 
      at processImmediate [as _immediateCallback] (timers.js:745:5) 
  From previous event: 
      at Object.<anonymous> (/usr/local/lib/node_modules/serverless/bin/serverless.js:61:4) 
      at Module._compile (module.js:653:30) 
      at Object.Module._extensions..js (module.js:664:10) 
      at Module.load (module.js:566:32) 
      at tryModuleLoad (module.js:506:12) 
      at Function.Module._load (module.js:498:3) 
      at Function.Module.runMain (module.js:694:10) 
      at startup (bootstrap_node.js:204:16) 
      at bootstrap_node.js:625:3 
  
  Get Support -------------------------------------------- 
     Docs:          docs.serverless.com 
     Bugs:          github.com/serverless/serverless/issues 
     Issues:        forum.serverless.com 
  
  Your Environment Information --------------------------- 
     Operating System:          linux 
     Node Version:              8.16.0 
     Framework Version:         1.57.0 
     Plugin Version:            3.2.3 
     SDK Version:               2.2.1 
     Components Core Version:   1.1.2 
     Components CLI Version:    1.4.0 
  
 
[Container] 2019/11/18 16:23:54 Command did not exit successfully cd core/ && bash deploy.sh exit status 1 
[Container] 2019/11/18 16:23:54 Phase complete: BUILD State: FAILED 
[Container] 2019/11/18 16:23:54 Phase context status code: COMMAND_EXECUTION_ERROR Message: Error while executing command: cd core/ && bash deploy.sh. Reason: exit status 1 
```

**Comment by ferasallaou** [comment]
Also, this happens when I activate Layers in `serverless.yml` it searches for nodeModules.zip inside `.serverless` even though I used `serverless deploy --package deployment_package --stage development`

**Comment by medikoo** [comment]
@ferasallaou thank you, I see it clearly now. Fix will be introduced shortly

**Comment by soelchma** [comment]
> @ferasallaou thank you, I see it clearly now. Fix will be introduced shortly

**Comment by soelchma** [comment]
@medikoo and @ferasallaou thanks guys and sorry for coming back so late!

**Comment by ferasallaou** [comment]
@soelchma is there any expected date to solve this? I think it is a priority, I cannot complete Pipeline deployments because of this :(

**Comment by medikoo** [comment]
> @soelchma is there any expected date to solve this?

Hopefully patch will land next week

**Comment by ferasallaou** [comment]
@medikoo Any updates on this one? :(

**Comment by medikoo** [comment]
@ferasallaou I'm really sorry, other priorities took over this priority. Fix will definitely land (work on it has started), but at this point I cannot promise any specific release window

**Comment by soelchma** [comment]
@medikoo Ok thank you!

**Comment by danyfu** [comment]
Hi a workaround solution I found.
Instead of using `$CODEBUILD_SRC_DIR` variable package using a relative path.

So you can change your serverless deploy to:
`sls deploy --stage $env --package ./target/$env -v -r eu-central-1`

**Comment by franciscomemoli** [comment]
> Hi a workaround solution I found.
> Instead of using `$CODEBUILD_SRC_DIR` variable for the path of package I changed using a relative path.
> 
> So you can change your serverless deploy to:
> `sls deploy --stage $env --package ./target/$env -v -r eu-central-1`

didn't work for me

**Comment by debugguru** [comment]
Any update on this? This is still an issue

**Comment by jplandry908** [comment]
For anyone else experiencing this issue, I finally identified the root cause and a simple workaround.

In summary, the **"artifact"** incorrectly has **"$CODEBUILD_SRC_DIR"** in the value. Also, the _srcxxxxxxxxxx_ number changes between deploy and build, so you can't just do a find/replace for the exact string in DEPLOY since the value is the $CODEBUILD_SRC_DIR that was in BUILD.

For each function, you will see an artifact property that looks similar to:
`"artifact": "/codebuild/output/src887072543/src/.serverless/function.zip"`
while it should actually look like this:
`"artifact": ".serverless/function.zip"`

To work around this issue, I simply added the following two lines in my BUILD commands, immediately above the '**serverless deploy ...**' statement.

`- sed -i 's/\/codebuild\/output//g' artifacts/${!stage}/serverless-state.json`
`- sed -i 's/\/src.*\//.serverless\//g' artifacts/${!stage}/serverless-state.json`

This will transform the "/codebuild/output/src887072543/src/.serverless/function.zip" statements to ".serverless/function.zip", allowing the deploy process to find the file without error.

This took me a while to figure out, so I hope it helps others that may be experiencing the same issue.

**Comment by czubocha** [comment]
Hey everyone, I just wanted to let you know that this issue should now be resolved in [v4.4.17](https://github.com/serverless/serverless/releases/tag/v4.4.17). If you encounter any further issues, feel free to reach out!
